url04f7fcf4e1_peteratt_HDFS_Replication_Simulator_tgz-pJ8-hdfs_replicationsimulator_MainJ8/src/hdfs/replicationsimulator/Namenode.java

```diff
@@ -0,0 +1,654 @@
+		pendingReplications = new PendingReplicationBlocks();
+		hbthread.start();
+		replthread.start();
+	}
+
+	/**
+	 * Periodically calls heartbeatCheck().
+	 */
+	class HeartbeatMonitor implements Runnable {
+		/**
+	   */
+		public void run() {
+			Simulator.addTrace(new SimTrace("Namenode heartbeat monitor created"));
+
+			while (isRunning) {
+				try {
+					// TODO
+					heartbeatCheck();
+					Thread.sleep(heartbeatRecheckInterval);
+				} catch (Exception e) {
+					// FSNamesystem.LOG.error(StringUtils.stringifyException(e));
+				}
+			}
+		}
+	}
+
+	/**
+	 * Check if there are any expired heartbeats, and if so, whether any blocks
+	 * have to be re-replicated. While removing dead datanodes, make sure that
+	 * only one datanode is marked dead at a time within the synchronized
+	 * section. Otherwise, a cascading effect causes more datanodes to be
+	 * declared dead.
+	 */
+	void heartbeatCheck() {
+		boolean allAlive = false;
+		while (!allAlive) {
+			boolean foundDead = false;
+			int nodeID = -1;
+
+			// locate the first dead node.
+			synchronized (heartbeats) {
+				for (Iterator<DatanodeInfo> it = heartbeats.iterator(); it
+						.hasNext();) {
+					DatanodeInfo nodeInfo = it.next();
+					if (isDatanodeDead(nodeInfo)) {
+						foundDead = true;
+						nodeID = nodeInfo.getId();
+						break;
+					}
+				}
+			}
+
+			// acquire the fsnamesystem lock, and then remove the dead node.
+			if (foundDead) {
+				synchronized (this) {
+
+					synchronized (DatanodesMap) {
+						DatanodeInfo nodeInfo = null;
+						nodeInfo = DatanodesMap.getDatanodeInfo(nodeID);
+
+						if (nodeInfo != null && isDatanodeDead(nodeInfo)) {
+							/*
+							 * NameNode.stateChangeLog.info(
+							 * "BLOCK* NameSystem.heartbeatCheck: " +
+							 * "lost heartbeat from " + nodeInfo.getName());
+							 */
+							Simulator.addTrace(new SimTrace(
+									SimTrace.FAILURE_DETECTION, nodeInfo
+											.getId()));
+							removeDatanode(nodeInfo);
+
+						}
+					}
+				}
+
+			}
+			allAlive = !foundDead;
+		}
+	}
+
+	private void removeDatanode(DatanodeInfo nodeInfo) {
+		DatanodesMap.removeDatanode(nodeInfo);
+
+		/*
+		for (Iterator<BlockInfo> it = nodeInfo.getBlockIterator(); it.hasNext();) {
+			removeStoredBlock(it.next(), nodeInfo);
+		}
+		*/
+		for(BlockInfo block: nodeInfo.getBlockList()){
+			removeStoredBlock(block,nodeInfo);
+		}
+
+	}
+
+	private boolean isDatanodeDead(DatanodeInfo node) {
+		return (node.getLastHB() < (Node.now() - heartbeatExpireInterval));
+	}
+
+	/**
+	 * Periodically calls computeReplicationWork().
+	 */
+	class ReplicationMonitor implements Runnable {
+		static final int INVALIDATE_WORK_PCT_PER_ITERATION = 32;
+		static final float REPLICATION_WORK_MULTIPLIER_PER_ITERATION = 2;
+
+		public void run() {
+			while (isRunning) {
+
+					computeDatanodeWork();
+					processPendingReplications();
+					try {
+						Thread.sleep(replicationRecheckInterval);
+					} catch (InterruptedException e) {
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem**: InterruptedException is ignored. This can delay thread shutdown and clear the thread’s interrupt status. Only code that implements a thread’s interruption policy can swallow an interruption request.

**Fix**: Rethrow the InterruptedException or reinterrupt the current thread using *Thread.currentThread().interrupt()* so that higher-level interrupt handlers can function correctly.
 If you are wrapping the InterruptedException inside a RuntimeException, call *Thread.currentThread().interrupt()* before throwing the RuntimeException.

Learn more about [interrupts](https://docs.oracle.com/javase/tutorial/essential/concurrency/interrupt.html) and [dealing with InterruptedException](https://www.ibm.com/developerworks/java/library/j-jtp05236/index.html?ca=drs-#N10187)